File: web/scripts/build-icons.php
<?php

$base = dirname(__DIR__);
$icons = json_decode(file_get_contents($base . '/icons.json'), true);
$template_dir = $base . '/../theme/src';
$output_dir = $base . '/icons';

$templates = glob($template_dir . '/template_*.svg');
if (!$templates) {
    fprintf(STDERR, "error: no templates found in %s\n", $template_dir);
    exit(1);
}

$count = 0;

foreach ($templates as $tpl_path) {
    $variant = preg_replace('/^template_/', '', pathinfo($tpl_path, PATHINFO_FILENAME));
    $raw = file_get_contents($tpl_path);

    $optimized = preg_replace('/<\?xml[^?]*\?>\s*/', '', $raw);
    $optimized = preg_replace('/\s*xmlns:dc="[^"]*"/', '', $optimized);
    $optimized = preg_replace('/\s*xmlns:cc="[^"]*"/', '', $optimized);
    $optimized = preg_replace('/\s*xmlns:rdf="[^"]*"/', '', $optimized);
    $optimized = preg_replace('/\s*xmlns:svg="[^"]*"/', '', $optimized);
    $optimized = preg_replace('/\s*xmlns:xlink="[^"]*"/', '', $optimized);
    $optimized = str_replace('xlink:href', 'href', $optimized);
    $optimized = preg_replace('/\s*version="[^"]*"/', '', $optimized);
    $optimized = preg_replace('/\s*<desc>[^<]*<\/desc>/', '', $optimized);
    $optimized = preg_replace('/\s*<metadata>.*?<\/metadata>/s', '', $optimized);
    $optimized = preg_replace('/\s*id="(start|stop)"/', '', $optimized);
    $optimized = preg_replace('/;stop-opacity:1/', '', $optimized);
    $optimized = preg_replace('/;fill-opacity:1;stroke:none/', '', $optimized);
    $optimized = preg_replace('/\n\s*\n/', "\n", $optimized);

    $dir = $output_dir . '/' . $variant;
    if (!is_dir($dir)) {
        mkdir($dir, 0755, true);
    }

    foreach ($icons as $icon) {
        $title = ucfirst(str_replace('-', ' ', $icon['name']));
        $svg = str_replace(['{{TITLE}}', '{{PATH}}'], [$title, $icon['d']], $optimized);
        file_put_contents($dir . '/' . $icon['name'] . '.svg', $svg);
        $count++;
    }
}

$variant_count = count($templates);
$icon_count = count($icons);
fprintf(STDERR, "built %d SVG files (%d variants × %d icons) in %s\n", $count, $variant_count, $icon_count, $output_dir);

File: web/scripts/build-icons.py
#!/usr/bin/env python3

import json
import re
import sys
from pathlib import Path

base = Path(__file__).resolve().parent.parent
icons = json.loads((base / "icons.json").read_text())
template_dir = base.parent / "theme" / "src"
output_dir = base / "icons"

templates = sorted(template_dir.glob("template_*.svg"))
if not templates:
    print(f"error: no templates found in {template_dir}", file=sys.stderr)
    sys.exit(1)

STRIP = [
    (re.compile(r'<\?xml[^?]*\?>\s*'), ''),
    (re.compile(r'\s*xmlns:dc="[^"]*"'), ''),
    (re.compile(r'\s*xmlns:cc="[^"]*"'), ''),
    (re.compile(r'\s*xmlns:rdf="[^"]*"'), ''),
    (re.compile(r'\s*xmlns:svg="[^"]*"'), ''),
    (re.compile(r'\s*xmlns:xlink="[^"]*"'), ''),
    (re.compile(r'\s*version="[^"]*"'), ''),
    (re.compile(r'\s*<desc>[^<]*</desc>'), ''),
    (re.compile(r'\s*<metadata>.*?</metadata>', re.DOTALL), ''),
    (re.compile(r'\s*id="(?:start|stop)"'), ''),
    (re.compile(r';stop-opacity:1'), ''),
    (re.compile(r';fill-opacity:1;stroke:none'), ''),
    (re.compile(r'\n\s*\n'), '\n'),
]

count = 0

for tpl_path in templates:
    variant = tpl_path.stem.removeprefix("template_")
    raw = tpl_path.read_text()

    optimized = raw.replace("xlink:href", "href")
    for pattern, replacement in STRIP:
        optimized = pattern.sub(replacement, optimized)

    variant_dir = output_dir / variant
    variant_dir.mkdir(parents=True, exist_ok=True)

    for icon in icons:
        title = icon["name"].replace("-", " ").title()
        svg = optimized.replace("{{TITLE}}", title).replace("{{PATH}}", icon["d"])
        (variant_dir / f"{icon['name']}.svg").write_text(svg)
        count += 1

print(
    f"built {count} SVG files ({len(templates)} variants × {len(icons)} icons) in {output_dir}",
    file=sys.stderr,
)
